pyanaconda/bootloader.py

```diff
@@ -1677,6 +1677,119 @@ def check(self):
+
+        return exec_func("efibootmgr", list(args), **kwargs)
+
+    #
+    # configuration
+    #
+
+    @property
+    def efi_product_path(self):
+        """ The EFI product path.
+
+            eg: HD(1,800,64000,faacb4ef-e361-455e-bd97-ca33632550c3)
+        """
+        buf = self.efibootmgr("-v", stderr="/dev/tty5", capture=True)
+        matches = re.search(productName + r'\s+(HD\(.+?\))', buf)
+        if matches and matches.groups():
+            return matches.group(1)
+        return ""
+
+    @property
+    def grub_conf_device_line(self):
+        return "device %s %s\n" % (self.grub_device_name(self.stage2_device),
+                                   self.efi_product_path)
+
+    #
+    # installation
+    #
+
+    def remove_efi_boot_target(self):
+        buf = self.efibootmgr(capture=True)
+        for line in buf.splitlines():
+            try:
+                (slot, _product) = line.split(None, 1)
+            except ValueError:
+                continue
+
+            if _product == productName:
+                slot_id = slot[4:8]
+                # slot_id is hex, we can't use .isint and use this regex:
+                if not re.match("^[0-9a-fA-F]+$", slot_id):
+                    log.warning("failed to parse efi boot slot (%s)", slot)
+                    continue
+
+                rc = self.efibootmgr("-b", slot_id, "-B",
+                                     root=iutil.getSysroot())
+                if rc:
+                    raise BootLoaderError("failed to remove old efi boot entry")
+
+    def _add_single_efi_boot_target(self, partition):
+        boot_disk = partition.disk
+        boot_part_num = str(partition.partedPartition.number)
+
+        rc = self.efibootmgr("-c", "-w", "-L", productName,
+                             "-d", boot_disk.path, "-p", boot_part_num,
+                             "-l", "\\EFI\\redhat\\grub.efi",
+                             root=iutil.getSysroot())
+        if rc:
+            raise BootLoaderError("failed to set new efi boot target. This is most likely a kernel or firmware bug.")
+
+    def add_efi_boot_target(self):
+        if self.stage1_device.type == "partition":
+            self._add_single_efi_boot_target(self.stage1_device)
+        elif self.stage1_device.type == "mdarray":
+            for parent in self.stage1_device.parents:
+                self._add_single_efi_boot_target(parent)
+
+    def install(self, args=None):
+        self.remove_efi_boot_target()
+        self.add_efi_boot_target()
+
+    def update(self):
+        self.install()
+
+    #
+    # installation
+    #
+    def write(self):
+        """ Write the bootloader configuration and install the bootloader. """
+        if self.update_only:
+            self.update()
+            return
+
+        os.sync()
+        self.stage2_device.format.sync(root=iutil.getTargetPhysicalRoot())
+        self.install()
+        self.write_config()
+
 class EFIGRUB(GRUB2):
```

Comment:
I can't believe I'm going to say this. But I think we need a mixin class for EFI, we shouldn't be duplicating the code between EFIGRUB and EFIGRUB1. Take all the common parts and put it in there and then use class EFIGRUB(GRUB2, EFIMixin) or whatever the right name is for it.